### PR TITLE
SALTO-6955: Move local-workspace files to a new OSS package

### PR DESCRIPTION
**Pull request by dorsev**
### TODO
- [x] Self-review this code.
- [x] Validate next prs work properly rebased on this PR

### Summary
This PR introduces the `local-workspace` package, which handles operations related to local workspace management, such as state management, file storage, configuration, and telemetry.

It will be utilized in a follow-up PR [#6868](https://github.com/salto-io/salto/pull/6868) to implement adapter-format methods for the dummy-adapter. This was previously infeasible due to a cyclic dependency.

---

### Additional Context for Reviewers
I’ve ensured that no changes break the `core` API package. Please double-check that nothing was missed.

---

### Release Notes
None.

---

### User Notifications
None.

### CODE REVIEW

File: packages/core/index.ts [hunk 1]
```diff
@@ -23,23 +23,14 @@ export {
 export { createDiffChanges, getEnvsDeletionsDiff } from './src/core/diff'
 export { RenameElementIdError } from './src/core/rename'
 export {
-  loadLocalWorkspace,
```
Comment (dorsev) [inline packages/core/index.ts:26]: Note: Ive made sure all exports before exist after the change

Comment (dorsev) [reply packages/core/index.ts:26]: Feel free to doublecheck me on this.

File: packages/core/src/local-workspace/workspace.ts [hunk 2]
```diff
@@ -283,125 +84,23 @@ type LoadLocalWorkspaceArgs = {
...
-  const ws = await loadWorkspace(
-    workspaceConfigSrc,
-    adaptersConfig,
-    credentials,
-    elemSources,
-    remoteMapCreator,
-    ignoreFileChanges,
-    persistent,
-    undefined,
+export async function loadLocalWorkspace(args: LoadLocalWorkspaceArgs): Promise<Workspace> {
```
Comment (dorsev) [inline packages/core/src/local-workspace/workspace.ts:87]: To view the diff against `workspace.ts` in local-workspace package, see this gist
https://gist.github.com/dorsev/7f0546920f1f68dd2447570c39581bc6/revisions

File: packages/local-workspace/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Salto: Local Workspace
+
+This project is the Salto Local workspace, responsible for managing local workspace operations including state management, file storage, configuration handling, and telemetry collection
```
Comment (dorsev) [inline packages/local-workspace/README.md:3]: @ori-moisis  - is this accurate enough?

File: packages/local-workspace/src/workspace.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+
+type GetNaclFilesSourceParamsArgs = {
+  sourceBaseDir: string
+  name: string
+  remoteMapCreator: remoteMap.RemoteMapCreator
+  persistent: boolean
+  excludeDirs?: string[]
+}
+
+const getNaclFilesSourceParams = ({
```
Comment (dorsev) [inline packages/local-workspace/src/workspace.ts:79]: To compare this file against the workspace/workspace.ts
see - https://gist.github.com/dorsev/7f0546920f1f68dd2447570c39581bc6

File: packages/local-workspace/test/workspace.test.ts [hunk 5]
```diff
@@ -478,89 +515,4 @@ describe('local workspace', () => {
       expect(file.rm).not.toHaveBeenCalled()
     })
   })
-
-  describe('getCustomReferences', () => {
```
Comment (dorsev) [inline packages/local-workspace/test/workspace.test.ts:482]: These stayed at packages/core/test/local-workspace/workspace.test.ts

File: packages/workspace/index.ts [hunk 6]
```diff
@@ -86,6 +87,7 @@ export {
   flags,
   elementSource,
   remoteMap,
+  WorkspaceGetCustomReferencesFunc,
```
Comment (dorsev) [inline packages/workspace/index.ts:90]: This is needed for local-workspace types

**Comment by dorsev** [comment]
`ci/circleci: test-utils — Your tests failed on CircleCI` failed on `Upload  Coverage result to coveralls`
Will re-run

File: packages/local-workspace/generate.sh
Comment (ori-moisis) [inline packages/local-workspace/generate.sh:1]: we can remove this file
we can remove it from `core` as well (don't forget to remove the command from `package.json`)

File: packages/local-workspace/jest.config.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    '!**/hcl.ts', // Generated parser file
+    '!<rootDir>/index.ts',
+  ],
+  coverageThreshold: {
+    // Slowly start increasing here, never decrease!
+    global: {
+      branches: 83.22,
+      functions: 88.79,
+      lines: 85,
+      statements: 90,
```
Comment (ori-moisis) [inline packages/local-workspace/jest.config.js:23]: can we raise these?

File: packages/local-workspace/jest.config.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * CERTAIN THIRD PARTY SOFTWARE MAY BE CONTAINED IN PORTIONS OF THE SOFTWARE. See NOTICE FILE AT https://github.com/salto-io/salto/blob/main/NOTICES
+ */
+const deepMerge = require('../../build_utils/deep_merge')
+
+module.exports = deepMerge(require('../../jest.base.config.js'), {
+  displayName: 'salto',
+  rootDir: `${__dirname}`,
+  collectCoverageFrom: [
+    '!**/hcl.ts', // Generated parser file
```
Comment (ori-moisis) [inline packages/local-workspace/jest.config.js:14]: probably don't need this line

File: packages/local-workspace/package.json [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    "test": "jest",
+    "clean": "rm -rf ./dist .eslintcache  ./src/generated",
+    "watch-test": "yarn tsc-watch",
+    "build-ts": "tsc -b",
+    "watch-ts": "tsc -b -w",
+    "lint": "eslint --max-warnings 0 ./",
+    "lint-fix": "yarn lint --fix",
+    "format": "prettier --write . --ignore-path=../../.prettierignore --ignore-path=../../.gitignore --ignore-path=.gitignore",
+    "check-format": "prettier --check . --ignore-path=../../.prettierignore --ignore-path=../../.gitignore --ignore-path=.gitignore",
+    "generate": "./generate.sh"
```
Comment (ori-moisis) [inline packages/local-workspace/package.json:30]: remove this

File: packages/local-workspace/package.json [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  "files": [
+    "dist/src",
+    "dist/index.*"
+  ],
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "build": "../../build_utils/turbo_run.sh build-ts ; ../../build_utils/turbo_run.sh lint",
+    "test": "jest",
+    "clean": "rm -rf ./dist .eslintcache  ./src/generated",
```
Comment (ori-moisis) [inline packages/local-workspace/package.json:22]: remove the `./src/generated`?

File: packages/local-workspace/src/flags.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright 2024 Salto Labs Ltd.
+ * Licensed under the Salto Terms of Use (the "License");
+ * You may not use this file except in compliance with the License.  You may obtain a copy of the License at https://www.salto.io/terms-of-use
+ *
+ * CERTAIN THIRD PARTY SOFTWARE MAY BE CONTAINED IN PORTIONS OF THE SOFTWARE. See NOTICE FILE AT https://github.com/salto-io/salto/blob/main/NOTICES
+ */
+
+export const LOCAL_WORKSPACE_FLAGS = {
+  dumpStateWithLegacyFormat: 'DUMP_STATE_WITH_LEGACY_FORMAT',
```
Comment (ori-moisis) [inline packages/local-workspace/src/flags.ts:10]: should be removed from core flags

File: packages/local-workspace/test/common/helpers.ts
Comment (ori-moisis) [inline packages/local-workspace/test/common/helpers.ts:1]: I think this should not all be copied.
IIUC, some of the functions are needed only here (can be removed from `core`, e.g. `inMemRemoteMapCreator`), some of them are not needed here at all (e.g. `createMockAdapter`)

worth going over the other files here as well

**Review by ori-moisis — APPROVED** [review]

File: packages/local-workspace/jest.config.js (hunk 7 above)
Comment (dorsev) [reply packages/local-workspace/jest.config.js:23]: raised

**Pull request merged**